### PR TITLE
fix: Show correct headsigns and groupings on all GL branches

### PR DESCRIPTION
**Pull request by EmmaSimon**
### Summary

_Ticket:_ [GL showing duplicate Eastbound @ Gov Center with no data](https://app.asana.com/0/1205732265579288/1208378317839446/f)

Most of the added lines here are in the tests, sorry for the large PR! 😅 

After #454, there were issues introduced where we would only ever display grouped nearby headsigns on the Medford/Tufts branch. This changes the headsign grouping to split grouped routes into separate headsigns if only a single route pattern is predicted in one direction. This means that it's technically possible (but very unlikely) for a grouping to be displayed when an atypical trip is predicted but has multiple other predictions before it and so wouldn't be displayed in nearby, but since this only applies to a stretch of 5 stops that rarely have more than 2 trains headed inbound at any time, this shouldn't happen often.

This also includes changes to use the pattern headsign to generate directions, since the destination names in the route only apply to the route's typical pattern, and includes a default null `Direction` object in the `StaticPatterns.ByHeadsign`, because we still want the special cased `Direction` object (like "Park St & North") used in `groupedDirection` to set the overall directions in `StopPatterns.ForLine` for displaying in the filtered stop details direction toggle.

Screenshots of correct behavior after these changes:

Simulator Screenshot - iPhone 15 Pro - 2024-10-15 at 11 42 17Simulator Screenshot - iPhone 15 Pro - 2024-10-15 at 11 42 22Simulator Screenshot - iPhone 15 Pro - 2024-10-15 at 11 46 04Simulator Screenshot - iPhone 15 Pro - 2024-10-15 at 11 46 09

### Testing

Added tests for newly added functionality and updated Direction behavior.

### CODE REVIEW

File: shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt [hunk 1]
```diff
@@ -99,4 +101,143 @@ data class PatternsByStop(
             }
             .distinct()
     }
+
+    companion object {
+        // Even if a direction can serve multiple routes according to the static data, it's possible
+        // that only one of those routes is typical, in which case we don't want to display it as a
+        // grouped direction in the UI. If there are only predicted trips on a single route, this
+        // will split the grouped direction up into as many headsign rows as there are headsigns.
```
Comment (KaylaBrady) [inline shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt:109]: praise: this is very descriptive, thank you!

File: shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt [hunk 2]
```diff
@@ -99,4 +101,143 @@ data class PatternsByStop(
...
+        fun resolveRealtimePatternForDirection(
```
Comment (KaylaBrady) [inline shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt:110]: suggestion: this function is pretty long - consider breaking out some of the helper steps into separate functions

File: shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt [hunk 3]
```diff
@@ -99,4 +101,143 @@ data class PatternsByStop(
...
+                staticData.routes.firstOrNull {
+                    it.id == headsignsAndPatternsToDisplayByRoute.keys.firstOrNull()
+                }
+            if (headsignsAndPatternsToDisplayByRoute.size == 1 && firstDisplayedRoute != null) {
+                val headsignsToDisplay =
+                    (headsignsAndPatternsToDisplayByRoute[firstDisplayedRoute.id] ?: emptyMap())
+                return headsignsToDisplay.map { (headsign, patternIds) ->
+                    val patternsForHeadsign = patternIds.mapNotNull { patternsById[it] }
+                    RealtimePatterns.ByHeadsign(
+                        NearbyStaticData.StaticPatterns.ByHeadsign(
```
Comment (KaylaBrady) [inline shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt:215]: question (non-blocking): would it be possible / simplify things at all to use the `RealtimePatterns.ByHeadsign` constructor that doesn't take a `StaticPatterns.ByHeadsign`?

File: shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt [hunk 4]
```diff
@@ -99,4 +101,143 @@ data class PatternsByStop(
...
+                    .toMap()
+            val upcomingPatternsByRouteAndHeadsign =
+                tripsByPattern.values
+                    .flatten()
+                    .groupBy({ it.trip.routeId }, { it.trip.headsign to it.trip.routePatternId })
+                    .mapValues { routeEntry ->
+                        routeEntry.value.groupBy({ it.first }, { it.second })
+                    }
+
+            val predictedPatternsByRoute =
```
Comment (KaylaBrady) [inline shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt:166]: question (non-blocking): Confirming my own learning - if we skip the step of checking predicted patterns by route, then we'd break down into showing individual headsigns prematurely because there could be some late-at-night atypical headsigns.

File: shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt [hunk 5]
```diff
@@ -99,4 +101,143 @@ data class PatternsByStop(
...
+                staticData.patterns
+                    .filter {
+                        tripsByPattern[it.id]?.any { trip ->
+                            trip.prediction != null && trip.vehicle?.tripId == trip.trip.id
```
Comment (KaylaBrady) [inline shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt:170]: question: why check the vehicle's trip ID here? Wouldn't we still show the prediction even if the vehicle is on another trip?

File: shared/src/commonTest/kotlin/com/mbta/tid/mbta_app/model/PatternsByStopTest.kt [hunk 6]
```diff
@@ -300,4 +300,309 @@ class PatternsByStopTest {
         assertEquals(listOf(alert), patternsByStop.alertsHereFor(0, global))
         assertEquals(listOf(alert), patternsByStop.alertsHereFor(1, global))
     }
+
+    @Test
+    fun `resolveRealtimePatternForDirection splits direction groups into headsigns when needed`() {
```
Comment (KaylaBrady) [inline shared/src/commonTest/kotlin/com/mbta/tid/mbta_app/model/PatternsByStopTest.kt:305]: suggestion: I'm having a bit of a hard time understanding this test since there is a lot of data defined & different cases asserted. Consider breaking this down into smaller tests that potentially share relevant data

**Review by KaylaBrady — COMMENTED** [review]
Nice work, this is tricky! Mostly comments from trying to wrap my head around this GL edge case better

File: shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt (hunk 3 above)
Comment (EmmaSimon) [reply shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt:215]: I actually set it up like that initially, by adding a new constructor for `RealtimePatterns.ByHeadsign` that took a `StaticPatterns.ByDirection`, then decided that doing it like this was actually the most straightforward approach.

File: shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/NearbyStaticData.kt [hunk 7]
```diff
@@ -113,20 +115,32 @@ data class NearbyStaticData(val data: List<TransitWithStops>) {
...
+                    } else if (typicalHere.size > 1) {
+                        Direction(
+                            routes.first().directionNames[directionId] ?: "",
+                            null,
+                            directionId
+                        )
+                    } else {
+                        when (val it = typicalHere.first()) {
+                            is StaticPatterns.ByDirection -> it.direction
+                            is StaticPatterns.ByHeadsign -> it.direction
```
[the remark this answers is not in the document]
Comment (EmmaSimon) [reply shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/NearbyStaticData.kt:136]: That's right, yeah. I think that would add additional confusion to all of the other places we use this, and here we would still need to do something in the case where it was a regular `StaticPatterns.ByHeadsign`, even though that should never be possible.

File: shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt (hunk 5 above)
Comment (EmmaSimon) [reply shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt:170]: Hmm, maybe it's a bad idea. I wanted to prevent cases where we would be displaying the grouping without actually displaying the atypical trip. So like, if there was a prediction for a C which is currently approaching Medford Tufts to start a new trip, but two E trains in front of it heading west, we would still show the group even though only the E trains are shown in the upcoming predictions. But maybe that's fine and this is too cautious and will hide predictions when we don't want it to.

Comment (EmmaSimon) [reply shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt]: I removed this check, I think it'll be okay to briefly show groups sometimes even when it's not totally clear why.

File: shared/src/commonTest/kotlin/com/mbta/tid/mbta_app/model/PatternsByStopTest.kt (hunk 6 above)
Comment (EmmaSimon) [reply shared/src/commonTest/kotlin/com/mbta/tid/mbta_app/model/PatternsByStopTest.kt]: Thanks for suggesting this! I quite like the pattern I ended up with with all the shared data in an object that each case references, I'll try to remember to do that sort of thing going forward when multiple tests share similar data.

**Review by KaylaBrady — APPROVED** [review]
